lib/ClangImporter/ClangImporter.cpp

```diff
@@ -2256,6 +2256,120 @@ ClangModuleUnit *ClangImporter::Implementation::getClangModuleForDecl(
   return getWrapperForModule(M);
 }
 
+// TODO: Consolidate with importSourceLoc below
+SourceLoc ClangImporter::Implementation::convertSourceLocation(
+    const clang::SourceLocation &loc) {
+  return getBufferImporterForDiagnostics().resolveSourceLocation(
+      getClangASTContext().getSourceManager(), loc);
+}
+
+bool ClangImporter::Implementation::setDiagnosticTargetIfNotAlreadyReported(
+    const SwiftLookupTable::SingleEntry &target) {
+  bool shouldReset = false;
+  bool alreadyReported = reportedDiagnosticTargets.count(target);
+  if (!alreadyReported && !getEagerImportActive()) {
+    shouldReset = setDiagnosticTarget(target);
+    reportedDiagnosticTargets.insert(target);
+  }
+  return !alreadyReported && !getEagerImportActive() && shouldReset;
+}
+
+bool ClangImporter::Implementation::setDiagnosticTarget(
+    const SwiftLookupTable::SingleEntry &target) {
+  assert(target);
+  // In some situations, the ClangImporter is invoked in a reentrant manner,
+  // resulting in a diagnosticTarget already being set. In such a case, we
+  // keep the outer diagnostic target.
+  if (!diagnosticTarget) {
+    diagnosticTarget = target;
+    return true;
+  }
+  return false;
+}
+
+SwiftLookupTable::SingleEntry
+ClangImporter::Implementation::getDiagnosticTarget() {
+  return diagnosticTarget;
+}
+
+void ClangImporter::Implementation::resetDiagnosticTarget(bool shouldReset) {
+  if (shouldReset) {
+    assert(diagnosticTarget);
+    diagnosticTarget = nullptr;
+  }
+}
+
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const SwiftLookupTable::SingleEntry &targetCandidate) const {
+  return diagnosticTarget == targetCandidate;
+}
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const clang::NamedDecl *targetCandidate) const {
+  return diagnosticTarget.dyn_cast<clang::NamedDecl *>() == targetCandidate;
+}
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const clang::ModuleMacro *targetCandidate) const {
+  return diagnosticTarget.dyn_cast<clang::ModuleMacro *>() == targetCandidate;
+}
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const clang::MacroInfo *targetCandidate) const {
+  return diagnosticTarget.dyn_cast<clang::MacroInfo *>() == targetCandidate;
+}
+
+bool ClangImporter::Implementation::setEagerImportActive() {
+  bool shouldReset = !eagerImportActive;
+  eagerImportActive = true;
+  return shouldReset;
+}
+
+void ClangImporter::Implementation::resetEagerImportActive(bool shouldReset) {
+  if (shouldReset) {
+    eagerImportActive = false;
+  }
+}
```

Comment:
Yeah, I was thinking something like `llvm::SaveAndRestore`, though that is no more easy to follow than SWIFT_DEFER, if anything harder. 

It is not always used in SWIFT_DEFER, but when it is it is because there are many alternate returns and I don't want to litter the code with `resetEagerImportActive` calls for each return, which I think is sort of the point of SWIFT_DEFER.